### DISCUSSION TITLE
next dev watch all files?

### DISCUSSION THREAD

**Discussion by nickyhajal** [vercel/next.js#10725] [Help]
I have some apis in `/pages/api`. Those APIs pull in entity models from typeorm. 

If I change the route itself, things restart without any issue.

If I change the typeorm model, I have to manually kill next server and restart.

Is there anyway to get next dev to restart when any imported file is changed?

**Comment by timneutkens** [comment]
Can you create a minimal reproduction? The require cache is cleared with files that were changed, potentially the typeorm stuff falls outside of that (it might cache in a different way)

**Comment by lambrospetrou** [comment]
Maybe it's not the same underlying functionality but is there a way to specify additional directories to watch? I have my articles authored in Markdown files in a directory outside the `pages` therefore when I edit them they are not updated while running `next dev`, even with a full page reload. Is there a way to signal that any change in a given directory should clear the cache or something?

I use `getStaticProps` and `getStaticPaths` to generate the article page by reading the markdown files using the `fs`.

update: I just found [preview mode](https://nextjs.org/docs/advanced-features/preview-mode) which solves my problem.

**Comment by RobinCsl** [comment]
Hi :wave: 
I stumbled upon the same issue as @lambrospetrou above. Judging by their website's README, it seems they are manually reloading the browser whenever they make a change to their local markdown files.
I just found https://github.com/hashicorp/next-remote-watch and it seems to do the job of watching my `content` folder containing my markdown files and reloading the browser whenever I make a change. Given that the package uses undocumented Next.js APIs, it might break for any new version of Next.js but the experience is well worth it in my opinion.

**Reply by cyberwombat** [reply]
Currently broken and looks like it's abandoned. Pull requests for Next 13 are ignored

**Comment by clementoriol** [comment]
Would love an official solution for this :)

**Reply by timneutkens** [reply]
No reproduction was provided.

https://github.com/vercel/next.js/discussions/10725#discussioncomment-722

**Reply by lambrospetrou** [reply]
> No reproduction was provided.
> 
> [#10725 (comment)](https://github.com/vercel/next.js/discussions/10725#discussioncomment-722)

@clementoriol Are you referring to the original post or mine, because they are different issues after all. The Preview API solves the issue with static props, albeit somewhat inconvenient.

**Reply by clementoriol** [reply]
My own use-case is a bit different, we got image folders outside of public that we copy to public using the webpack copy plugin. So not a bug in my case, just adding bumping this to add that we would also benefit from the ability to watch outside folders.

**Reply by clementoriol** [reply]
@lambrospetrou Yeah I realise my answer was unclear, I was interested in an official solution for, as you said "specify additional directories to watch". But that may be off-topic here.

**Comment by noor-codes** [comment]
## I have the same issue with different files 
I have markdown files outside the `/pages` directory. I need to refresh the browser to cause a server-render process every time to get the latest changes in development.

If nextjs could track those files just like it tracks all `js` and  `ts` files. It would have been so nice.

**Comment by kevinkhill** [comment]
Why don't we all try and come up with a solution? I just started using nextjs and have run into this, I would like page reloads on `.md` file changes too....

**Comment by mikematos84** [comment]
I am doing the same. I have an outside folder `/posts` containing a series of `.mdx` files. It would be great to have an option like 

```js
export default {
  ...
  watchDirectories: ['./posts']
  ...
}
```
So that we can list other directories outside of the `/pages` directory to monitor for changes and force a refresh.

I was taking care of this with `chokidar` and referencing 

```js
app.server.hotReloading.send('building');
app.server.hotReloading.send('reloadPage');
```
and that worked well until I upgraded to the latest Next (13.4.4) and now it seems to be broken as those functions are no longer available as it now throws this error 

```sh
- error unhandledRejection: TypeError: Cannot read properties of undefined (reading 'hotReloader')
    at FSWatcher.<anonymous> (E:\Projects\Personal\next-test-app\server.js:55:16)
    at FSWatcher.emit (node:events:513:28)
    at FSWatcher.emitWithAll (E:\Projects\Personal\next-test-app\node_modules\chokidar\index.js:540:8)
    at FSWatcher._emit (E:\Projects\Personal\next-test-app\node_modules\chokidar\index.js:632:8)
    at listener (E:\Projects\Personal\next-test-app\node_modules\chokidar\lib\nodefs-handler.js:370:20)
```

**Comment by 0nitsoga** [comment]
**Proposed Workaround**: I have crafted a shell script that monitors the app directory. When a file within this directory is modified, the script "touches" another file known to be monitored by "vercel dev", thereby causing a rebuild. The key idea is to leverage the filesystem watch capability to indirectly inform vercel dev of changes it might not be aware of.

For optimal usage, ensure the script is active when making and saving changes. While the solution may appear a bit indirect, it integrates smoothly, especially when combined with a VSCode task, providing an almost seamless experience.


----------------------
**dev_toucher.sh**
```
#!/bin/bash

WATCHED_DIR="app/"
TOUCHED_FILE="api/app.py"

while inotifywait -r -e close_write "$WATCHED_DIR"; do
    touch "$TOUCHED_FILE"
done
```


----------------------
**task for tasks.json**
```
    {
        "label": "Vercel Dev",
        "type": "shell",
        "command": "./dev_toucher.sh & vercel dev",
        "group": {
            "isDefault": false
        }
    }
```
----------------------
Note: Ensure you have the necessary tools installed for the script to function, particularly inotify-tools.
